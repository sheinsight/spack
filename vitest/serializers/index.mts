/**
 * 统一的快照序列化器管理
 * 按优先级组合各种专用序列化器
 */

import type { SnapshotSerializer } from 'vitest';
import { pathNormalizerSerializer } from './path-normalizer.mts';
import { duplicateDependencySerializer } from './duplicate-dependency.mts';
import { errorsArraySerializer } from './errors-array.mts';
import { generalArraySerializer } from './general-array.mts';

/**
 * 序列化器优先级列表
 * 越前面的优先级越高
 */
const serializers: SnapshotSerializer[] = [
  pathNormalizerSerializer,  // 路径标准化应该最优先
  duplicateDependencySerializer,
  errorsArraySerializer,
  generalArraySerializer,
];

/**
 * 统一快照序列化器
 * 按优先级匹配并应用相应的序列化逻辑
 */
const unifiedSerializer: SnapshotSerializer = {
  test: (val: unknown): boolean => {
    // 检查是否有任何序列化器可以处理这个值
    return serializers.some((serializer) => serializer.test(val));
  },

  serialize: (val: any) => {
    // 按优先级应用序列化逻辑
    for (const serializer of serializers) {
      if (serializer.test(val)) {
        return (serializer as SnapshotSerializer).serialize(val);
      }
    }

    // 默认情况，不应该到达这里
    return JSON.stringify(val, null, 2);
  },
};

export default unifiedSerializer;
