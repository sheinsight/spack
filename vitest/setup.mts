process.env.NODE_ENV = 'test';
process.env.API_URL = 'http://localhost:3000';
process.env.RSPACK_BINDING = `${process.cwd()}/crates/binding`;
