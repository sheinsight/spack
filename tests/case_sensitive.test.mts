import { test, expect } from 'vitest';
import { experiments } from '@rspack/core';
import * as binding from '@shined/spack-binding';
import { runCompiler } from './test_case.mts';
import type { RawCaseSensitivePathsPluginOpts } from '@shined/spack-binding';

binding.registerCaseSensitivePathsPlugin();
const CaseSensitivePathsPlugin = experiments.createNativePlugin<
  [RawCaseSensitivePathsPluginOpts],
  RawCaseSensitivePathsPluginOpts
>(binding.CustomPluginNames.CaseSensitivePathsPlugin, (opt) => ({ ...opt }));

const plugin = new CaseSensitivePathsPlugin({});

test('should report errors when npm alias imports have incorrect case sensitivity', async () => {
  const result = await runCompiler({
    fixture: 'case_sensitive_npm_alias',
    plugins: [plugin],
  });

  expect(result.length).toBe(1);

  let message = result[0].message;

  expect(message).toContain(`Can't resolve`);
  expect(message).toContain(`rEact19`);
});

test('should report errors when file imports have incorrect case sensitivity', async () => {
  const result = await runCompiler({
    fixture: 'case_sensitive_local_file',
    plugins: [plugin],
  });
  expect(result.length).toBe(1);
  let message = result[0].message;
  expect(message).toContain(`Can't resolve`);
  expect(message).toContain(`./Hello`);
});
