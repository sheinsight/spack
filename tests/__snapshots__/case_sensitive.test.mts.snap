// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should report errors when file imports have incorrect case sensitivity 1`] = `
[
  {
    "message": "  × Can't resolve \\"./Hello\\" in \\"<ROOT>/tests/fixtures/case_sensitive_local_file/src/index.ts\\". (case mismatch)\\n   ╭─[<ROOT>/tests/fixtures/case_sensitive_local_file/src/index.ts:1:23]\\n 1 │ import { hello } from './Hello';\\n   ·                        ────┬────\\n   ·                            ╰── Path case mismatch\\n 2 │ \\n 3 │ hello();\\n   ╰────\\n  help: Fix the case of file paths to ensure consistency in cross-platform builds.\\n        It may work fine on macOS/Windows, but will fail on Linux.\\n",
    "code": "case mismatch",
    "moduleIdentifier": "<ROOT>/tests/fixtures/case_sensitive_local_file/src/index.ts",
    "moduleName": "./tests/fixtures/case_sensitive_local_file/src/index.ts",
    "moduleId": "./tests/fixtures/case_sensitive_local_file/src/index.ts",
    "moduleTrace": []
  }
]
`;

exports[`should report errors when npm alias imports have incorrect case sensitivity 1`] = `
[
  {
    "message": "  × Can't resolve \\"rEact19\\" in \\"<ROOT>/tests/fixtures/case_sensitive_npm_alias/src/index.ts\\". (case mismatch)\\n   ╭─[<ROOT>/tests/fixtures/case_sensitive_npm_alias/src/index.ts:1:19]\\n 1 │ import React from 'rEact19';\\n   ·                    ────┬────\\n   ·                        ╰── Path case mismatch\\n 2 │ \\n 3 │ console.log(React.version);\\n   ╰────\\n  help: Fix the case of file paths to ensure consistency in cross-platform builds.\\n        It may work fine on macOS/Windows, but will fail on Linux.\\n",
    "code": "case mismatch",
    "moduleIdentifier": "<ROOT>/tests/fixtures/case_sensitive_npm_alias/src/index.ts",
    "moduleName": "./tests/fixtures/case_sensitive_npm_alias/src/index.ts",
    "moduleId": "./tests/fixtures/case_sensitive_npm_alias/src/index.ts",
    "moduleTrace": []
  }
]
`;
